keps/sig-storage/1845-prioritization-on-volume-capacity/README.md

```diff
@@ -0,0 +1,424 @@
+
+## Motivation
+
+Because we didn't take volume capacity into account when scheduling a pod that
+can run in multiple topologies. A large PV may be used by a PVC with a small
+capacity request, even if there are many suitable small PVs in other
+topologies. PVCs with a large capacity request may not find feasible PVs to use
+if too many large PVs are consumed by PVCs with a small capacity request.
+
+Like CPU, Memory resources, the scheduler should take volume capacity into
+account in scheduling pods to ensure balanced resource usage.
+
+### Goals
+
+- Nodes of topology in which volumes with close capacity are preferred in
+  scheduling
+- Support arbitrary PV topology domain (e.g. hostname, zone, rack)
+
+### Non-Goals
+
+- Prioritizing nodes based on the total available capacity of static provisioned
+  PVs
+- Prioritizing nodes based on the total available number of static provisioned
+  PVs
+- Prioritizing nodes based on the total available capacity for dynamic
+  provisioning PVCs
+
+## Proposal
+
+### User Stories
+
+#### Volumes with close capacity are preferred
+
+As a cluster administrator, I need to prepare some local PVs with different
+storage capacities (e.g. 10Gi, 100Gi, 1Ti) as different workloads may have
+different storage requirements.
+
+A PV with 1Ti capacity on one node has the same chance as a PV on another node
+with 10Gi capacity to be bound with a 10Gi request PVC in scheduling. If nodes
+of topology in which volumes with close capacity are always preferred, we can
+achieve better storage resource usage.
+
+See https://github.com/kubernetes/kubernetes/issues/83323.
+
+#### Different weights for different classes
+
+Given a cluster which has two kinds of local storage classes:
+
+- hdd
+- ssd
+
+Workloads may request two disks, one from `hdd` and one from `ssd`. As
+the ssd storage resource may be relatively scarce, the administrator can
+configure a relatively larger weight for storage class `ssd` and hope it's more
+balanced.
+
+This applies to any other storage with topology constraints.
+
+## Design Details
+
+We have a scheduler plugin called
+[VolumeBinding](https://github.com/kubernetes/kubernetes/tree/master/pkg/scheduler/framework/plugins/volumebinding)
+to filter the nodes based on volume topology constraints. After the filter
+phase, there is a reduced set of nodes that can fit a pod. We can implement
+`Score` interfaces to score nodes based on assumed PVs on each node.
+
+The score is calculated in two steps:
+
+1) First, we group pod's unbound PVCs and corresponding static provisioned PVs
+by storage class. For PVCs/PVs in each storage class, we calculate requested to
+capacity ratio using the following formula:
+
+```
+[UtilizationPerClass] = [MaxUtilization] * ([RequestPerClass] / [CapacityPerClass])
```

Comment:
the volume scheduling library will choose the best fit in each topology (this is already done in our [PV search function](https://github.com/kubernetes/kubernetes/blob/db8a88721e587cbc54e7081cb88374e253cde4ed/pkg/controller/volume/persistentvolume/util/util.go#L293-L302)). sorry, I should add this.

in this case, the procedure will be:

- the filter phase
  - PV1-2 will be chosen for nodes in zone-1
  - PV2-1 will be chosen for nodes in zone-2
- in the scoring phase
  - nodes in zone-1 are preferred because they will get higher scores

Reply:
made an update https://github.com/kubernetes/enhancements/pull/1862/commits/ca712bfe6d8036996dc14c4228b8c0e7eb59b998

Reply:
Sounds good, thanks for explaining.